### PR TITLE
sql: add tests for UPSERT in the midst of schema changes

### PR DESCRIPTION
**Pull request by vivekmenezes**
Fix issues with the UPSERT implementation that were revealed
as a result of these tests.

I think we need to create a better API for the set of columns/indexes available for different operations. That's a refactor that we can do as part of a separate PR.

---

This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/9481)

### CODE REVIEW

**Comment by vivekmenezes** [comment]
@paperstreet any thoughts on these tests and fixes. Thanks

File: sql/descriptor_mutation_test.go [hunk 1]
```diff
@@ -642,6 +644,334 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR, i CHAR, INDEX foo (i, v), FAMIL
...
+			// updating "x" -> "w" results in the index updating from "x" -> "w",
+			// updating "z" -> "z" is a noop on the index.
+			_ = mTest.checkQueryResponse(indexQuery, [][]string{{"w"}, {"y"}, {"z"}})
+		}
+	}
+}
+
+// TestUpsertWithUniqueColumnMutation tests UPSERT operation while an index
+// mutation refers to a column mutation.
+func TestUpsertWithUniqueColumnMutation(t *testing.T) {
```
Comment (paperstreet) [inline sql/descriptor_mutation_test.go]: I'm finding it incredibly difficult to spot the differences between these 6 tests. Seems worth investigating another abstraction, something that (as much as possible) only leaves the bits that are different in the various test functions

File: sql/tablewriter.go [hunk 2]
```diff
@@ -271,6 +271,9 @@ func (tu *tableUpserter) init(txn *client.Txn) error {
 		if err != nil {
 			return err
 		}
+		// The fetchCols can also contain columns undergoing mutation; update
+		// requestCols.
+		requestedCols = tu.ru.fetchCols
```
Comment (paperstreet) [inline sql/tablewriter.go]: to match the rest of these, leave requestedCols alone here and iterate tu.ru.fetchCols below instead

File: sql/tablewriter.go [hunk 3]
```diff
@@ -279,15 +282,16 @@ func (tu *tableUpserter) init(txn *client.Txn) error {
...
+	for i, col := range requestedCols {
+		if _, ok := tu.fetchColIDtoRowIndex[col.ID]; ok {
 			valNeededForCol[i] = true
 		}
 	}
+
 	return tu.fetcher.Init(
 		tu.tableDesc, tu.fetchColIDtoRowIndex, &tu.tableDesc.PrimaryIndex, false, false,
-		tu.tableDesc.Columns, valNeededForCol)
+		requestedCols, valNeededForCol)
```
Comment (paperstreet) [inline sql/tablewriter.go]: s/requstedCols/tu.ru.fetchCols/

**Review by danhhz — COMMENTED** [review]
Thanks for catching this!

File: sql/descriptor_mutation_test.go [hunk 4]
```diff
@@ -493,8 +495,8 @@ CREATE TABLE t.test (k CHAR PRIMARY KEY, v CHAR, INDEX foo (v));
 	}
 }
 
-// TestOperationsWithUniqueColumnMutation tests all the operations while an
-// index mutation refers to a column mutation.
+// TestOperationsWithUniqueColumnMutation tests the INSERT, UPDATE and DELETE
+// operations while an index mutation refers to a column mutation.
 func TestOperationsWithUniqueColumnMutation(t *testing.T) {
```
Comment (vivekmenezes) [inline sql/descriptor_mutation_test.go]: done

Comment (vivekmenezes) [inline sql/descriptor_mutation_test.go]: done

File: sql/tablewriter.go (hunk 2 above)
Comment (vivekmenezes) [inline sql/tablewriter.go]: tu.ru is only set within this else clause. requestedCols is set above this if-then-else

**Comment by vivekmenezes** [comment]
@paperstreet I've consolidated UPSERT into the existing tests, but be warned github doesn't make it easy to review! It's ready for another review. Thanks!

File: sql/tablewriter.go (hunk 2 above)
Comment (paperstreet) [inline sql/tablewriter.go]: Ah, then add a fetchCols to tableUpserter and set it in both branches. In all of the "simple" Insert, Update, Delete tableWriters and rowWriters, the "requestedCols" is used as an immutable input and the "fetchCols" handed back is used to select what is fetched. I'd like to keep this one consistent

**Review by danhhz — APPROVED** [review]
LGTM modulo the requestedCols/fetchCols thing

**Pull request merged**